Add support for comparators to Sort.chpl

ben-albrecht opened the pull request:
## Added support for comparators to the Sort package + some other Sort module changes

### Comparators

In Chapel, comparators  are passed to the sort function as records with `.key(a)` or `.compare(a, b)` methods, rather than as functions themselves.

Some examples from the Sort.chpl *chpldoc-umentation*:

Example of `key(a)`:

```chapel

  var Array = [-1, -4, 2, 3];

  // Empty record serves as comparator
  record comparator { }

  // key method maps an element to the value to be used for comparison
  proc comparator.key(a) { return abs(a); }

  myComparator = new comparator();

  sort(Array, comparator=myComparator);

  // This will output: -1, 2, 3, -4
  writeln(Array);
```

Example of `compare(a, b)` sorting by absolute values:

```chapel

  var Array = [-1, -4, 2, 3];

  // Empty record serves as comparator
  record comparator { }

  // compare method defines how 2 elements are compared
  proc comparator.compare(a, b) {
      return abs(a) - abs(b);
  }

  myComparator = new comparator();

  sort(Array, comparator=myComparator);

  // This will output: -1, 2, 3, -4
  writeln(Array);
```

### Other stuff
* Added thorough chpldocumentation with examples
* Added tests for comparator support and expected errors (runtime and compile-time).
* Began deprecation of old sort function interface (supported `reverse` and `doublecheck` arguments + PascalCase instead of camelCase)
    * Threw `pragma "no doc"` on old sort functions
    * Staged deprecation warning as a comment
        * This will be uncommented in another PR in which we may decide to implement a standardized deprecation message function.
* Updated the base `chpl_compare` method to utilize comparators, and rely on the `defaultComparator` when no comparators are passed.
* Added convenient sort interface `sort()`, which currently acts as a wrapper for `quickSort`.
    * Updated standard library calls to `QuickSort` to `sort`

### Future Work
* Fully deprecate old sort interfaces
    *  Suppressable deprecation warnings for 1.14, then removed for 1.15
* Add comparator support to `selectionSort`
    * Work required with user-defined 
* Implement radix sort
* Possibly move individual sort algorithms to a SortAlgorithms module or submodule to reduce the complexity of Sort (since most uses of this library will only want to call 'sort', 'sorted' or 'isSorted').
* implement parallel sorting algorithms and have some reasonable way for users to request serial or parallel sorts.
* Create performance tracking tests and optimize/parallelize sort functions further
* Move Sort tests from `test/modules/standard` -> `test/modules/packages`


### TODO:

* [x] Add more complete testing (reverse, doublecheck, old version vs. new version)
* [x] Complete *chpldoc-umentation*
* [x] Rerun paratest
   * `[Summary: #Successes = 6260 | #Failures = 0 | #Futures = 0]`

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -18,27 +18,178 @@
+
+The ``key(a)`` method accepts one argument with the array ``eltType`` and
+returns a value that supports all of comparison operations: ``<``, ``<=``,
+``>``, ``>=``, or ``==``.
```
Are all of these actually used? E.g. you could implement them all with < ... if one didn't resolve.

mppf replied:
Actually I think it would be better if you said that the value returned by `key` must support all numeric operations, including comparison ( which you listed), and also things like - and <<.

I think it's more efficient to use - for our default comparator...
I'd use << and & for radix sorting...

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -18,27 +18,178 @@
+
+
+The ``compare(a, b)`` method accepts 2 arguments of the data ``eltType`` and
+returns an integer between ``1`` if ``a > b``, ``0`` if ``a == b``, and ``-1`` if ``a < b``.
```
I'd reword this comment. Also, generally speaking such comparison functions allow values >1 or < -1. See man qsort, which says:

        The comparison function must return an integer less than, equal to, or
     greater than zero if the first argument is considered to be respectively
     less than, equal to, or greater than the second.

I'd write:
The ``compare(a, b)`` method accepts 2 arguments of the data ``eltType`` and
returns an integer indicating how a and b compare to each other. It should return

  * > 0 if ``a > b``
  *    0 if ``a == b``
  * < 0 if ``a < b`

mppf replied:
By the way, the reason that it's not just -1 0 and 1 is so that you can write this:

    proc compare(a:int, b:int) {
      return a - b;
    }

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -18,27 +18,178 @@
+      return -1;
+    } else {
+      return 0;
+    }
```
I'd recommend you write this function body as

    return abs(a) - abs(b)

to emphasize and demonstrate the point about > 1 or < -1

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -18,27 +18,178 @@
+  writeln(Array);
+
+If both methods are implemented on the record passed as the comparator, the
+``compare(a, b)`` method will take priority over the ``key(a)`` method.
```
Hmm... interesting. Could you mark this as "current behavior" somehow? Once we add radix sort (which I think is ... key ... heh) it will not be able to use compare at all.

E.g.

In the current implementation, if both...

ben-albrecht replied:
We can swap the default priority for all sort functions. compare > key order was chosen somewhat arbitrarily.

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -18,27 +18,178 @@
+    else { return  comparator.compare(a, b); }
+  }
+
+  // TODO -- It may be faster to call key() on all data once, and mirror swaps
```
I'm fine with this comment, but Knuth volume 3 has a good discussion. You generally will only benefit from such a transformation if the values being sorted are much larger than their keys. I actually think that's less likely than the value being close to the same size as the key.

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -18,27 +18,178 @@
+  use Reflection;
+
+  // TODO -- we could also check types
+  // Use comparator.compare(a, b) if is defined by user
```
Do you know if Reflection works with type methods? I'm not so sure... I'd expect we'd need to beef up the Reflection module in order to check for a type method.

ben-albrecht replied:
Based on testing, they appear to be working fine.

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -18,27 +18,178 @@
+  // TODO -- It may be faster to call key() on all data once, and mirror swaps
+  // Compare results of comparator.key(a) if is defined by user
+  if canResolveMethod(comparator, "key", a) && canResolveMethod(comparator, "key", b) then
+    return chpl_compare(comparator.key(a), comparator.key(b), reverse);
```
See my comment above. I think this should return

if reverse then comparator.key(b) - comparator.key(a);
else comparator.key(a) - comparator.key(b);

I don't think we're getting much out of the `key` interface if it does not return a numeric value. I think people who have a "key" that is not numeric should just use the .compare interface.

mppf replied:
Of course you might disagree if you wanted .key to support strings, that's an interesting use...

ben-albrecht replied:
Your proposed change is much more elegant, but I would like to keep .key general if possible, e.g. supporting strings. However, a difference in performance might persuade me one way over the other.

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -18,27 +18,178 @@
-    if reverse then return a > b;
-    else return a < b;
+    if a < b { return -1; }
+    else if b < a { return 1; }
```
So, in the end you only used < if neither .key nor .compare were available? When does this happen? It probably needs mention in the module-level comment above.

mppf replied:
Or else this code can be removed?

ben-albrecht replied:
This is the default behavior when a comparator is not passed. If a user passes a comparator without a correctly defined `.key(a)` or `.compare(a,b)` method, they get a compile error.

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -18,27 +18,178 @@
+
+  if comparator.type == defaultComparator {}
+  else if canResolveMethod(comparator, "compare", a, a) {}
+  else if canResolveMethod(comparator, "key", a) {}
```
I think it would make sense to do some checks on procs available on the return type of key here as well.

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -47,24 +198,27 @@ private inline proc chpl_sort_cmp(a, b, param reverse=false, param eq=false) {
 
  */
-proc BubbleSort(Data: [?Dom] ?eltType, doublecheck=false, param reverse=false) where Dom.rank == 1 {
+proc BubbleSort(Data: [?Dom] ?eltType, doublecheck=false, param reverse=false, comparator:?rec=emptyRecord) where Dom.rank == 1 && isRecord(rec) {
```
I personally think that this `reverse` business is a waste of arguments.
Also doublecheck.

I can see keeping it for backwards compatibility, but I'd deprecate it.
For that reason I might prefer the actual sort algorithms be implemented with comparator and you just make other procs with the reverse argument that don't take in a comparator.
E.g. remove reverse and doublecheck from this one, and add
proc BubbleSort(Data, doublecheck=false, param reverse=false)
for backwards compatibility.

I'd probably criticize anybody who actually *calls* BubbleSort though. I'd personally advocate that we jettison it from the library completely.

But my comments will probably apply to the other sort variants.

mppf replied:
Also, I wouldn't put isRecord(rec) in the where clause. Let it be an error if chpl_check_comparator doesn't work. Nothing about how you are using the comparator needs it to be a record (it could be a class, or maybe somebody juts made methods on int, etc)

ben-albrecht replied:
* :+1: to deprecating `reverse` and `doublecheck` - Probably worth getting input from the rest of the team on this.
* I like bubble sort in that it's like the *hello world* of sorting. I feel like it's present more for completeness rather than actual usefulness.

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -163,46 +323,49 @@ proc InsertionSort(Data: [?Dom] ?eltType, doublecheck=false, param reverse=false
   }
 
-  for (a, _a) in zip(Data[lo..hi], _MergeIterator(A1, A2, reverse=reverse)) do a = _a;
+  for (a, _a) in zip(Data[lo..hi], _MergeIterator(A1, A2, reverse=reverse, comparator=comparator)) do a = _a;
```
I have trouble believing that this iterator-based implementation of MergeSort on arrays is anything other than a test of a language feature.
I'd expect MergeSort to outperform QuickSort if implemented well. I think this iterator is just adding overhead. It would be nice to add a comment indicating it's adding overhead (iterators with more than 1 yield can't be inlined) at not much benefit since the iterator is only called once.

ben-albrecht replied:
Agreed, but I'm going to leave this change for a separate PR for now.

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -263,14 +433,15 @@ proc QuickSort(Data: [?Dom] ?eltType, minlen=16, doublecheck=false, param revers
   Data(loptr) = pivotVal;
 
   //  cobegin {
```
Why is this Cobegin commented out?
I think it would be worthwhile to allow the user to choose whether the sort is parallel.

ben-albrecht replied:
Not sure, this is how I found it... I will leave this change to another PR as well.

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -306,11 +477,14 @@ proc SelectionSort(Data: [?Dom], doublecheck=false, param reverse=false) where D
 
  */
-inline proc VerifySort(Data: [?Dom] ?eltType, str: string, param reverse=false) {
+inline proc VerifySort(Data: [?Dom] ?eltType, str: string, param reverse=false, comparator:?rec=emptyRecord) where isRecord(rec) {
```
I'm very sure that this method does not need to be inline.

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -18,27 +18,178 @@
+``compare(a, b)`` method will take priority over the ``key(a)`` method.
 
  */
+
```
At the top-level, our Sort module is providing the wrong interface. We ask users to make choices that should be handled by our default.

For that reason, I'd recommend:
 * creating compatibility versions that use a special comparator and have existing signatures (and can be deprecated later)
 * removing reverse and  doublecheck arguments from the "real" versions of each sort algorithm
 * Also, for the "real" version of each, please rename it to not be upper case. This is one of the only places in our module where procs are not following the dominant style of starting with a lower case. That has always bothered me.
 * promoting VerifySort to a top-level function and rename it. I would call it isOrdered. As with the new sort routines, it should not take in a reverse argument - just a mapper. Make a deprecated version with the old signature if it's actually called in any tests.
 * Finally, we need to provide a way for users to get a parallel and non-parallel sort. We need a "Sort" method that chooses our best implementation. I would recommend we call this "sort" or "comparisonSort" and expect that later we will add "radixSort" which will have slightly different requirements on the comparator. I'd recommend using a non-param argument to indicate whether or not the sort should be parallel. A reasonable alternative would be to use a different method name for parallel/not parallel sorts.

ben-albrecht replied:
I will take care of all of these, except for the last point. I'd like to split that change off into a separate more broad "Sort Improvements" PR.

ben-albrecht commented:
@mppf - I think this is ready for final review now. Some comments:

* I've reverted the `compare` method back to the original basic implementation. I think we can optimize this in the future with overloaded version for different argument types.
* I've also decided to keep the `ReverseComparator` relatively simple for now. We can build upon this implementation for radixSort support in the future.
* The deprecated functions have been `pragma "no doc"`'d and I've added a commented-out deprecation message. I will un-comment this deprecation message and fix all the tests it breaks in a follow-up PR.
    * Maybe it's time to add a standardized deprecation message with a compiler flag for silencing the warning.
* `selectionSort` comparator support will also come in the near future. The challenge here lies in properly defining a comparator-supported reduction class.

mppf commented on modules/internal/DefaultAssociative.chpl:
```diff
@@ -407,7 +407,7 @@ module DefaultAssociative {
         tableCopy(tmp) = table[slot].idx;
   
-      QuickSort(tableCopy);
+      quickSort(tableCopy);
```
quickSort -> sort please

mppf commented on modules/internal/DefaultAssociative.chpl:
```diff
@@ -676,7 +676,7 @@ module DefaultAssociative {
         tableCopy(copy) = data(slot);
   
-      QuickSort(tableCopy);
+      quickSort(tableCopy);
```
here too

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -18,53 +18,321 @@
+
+  var absComparator: Comparator;
+
+  QuickSort(Array, comparator=absComparator);
```
QuickSort -> sort here too please

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -18,53 +18,321 @@
+
+  var absComparator: Comparator;
+
+  QuickSort(Array, comparator=absComparator);
```
QuickSort -> sort

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -18,53 +18,321 @@
+  var Array = [-1, -4, 2, 3];
+
+  // Using module-defined 'reversecomparator'
+  QuickSort(Array, comprator=reversecomparator)
```
QuickSort -> sort

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -18,53 +18,321 @@
+sorting is handled through the comparator interface.
+
+A module-defined :const:`reversecomparator` can be passed to a sort function to
+reverse the default sorting order.
```
I would have expected it to be named reverseComparator. Is there a reason you opted for the all-lower-case version?

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -18,53 +18,321 @@
+
+  var absReverseComparator: ReverseComparator(Comparator);
+
+  QuickSort(Array, comparator=absReverseComparator);
```
QuickSort -> sort

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -18,53 +18,321 @@
+  Instance of :record:`DefaultComparator` used as default ``comparator=``
+  argument when no comparator is passed to a sort function
+*/
+const defaultcomparator: DefaultComparator;
```
I think I would call this 'defaultOrder' or 'defaultComparator'

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -18,53 +18,321 @@
+   Instance of :record:`ReverseComparator`. Pass this as the ``comparator=``
+   argument of a sort function to reverse the sort order.
+ */
+const reversecomparator: ReverseComparator(DefaultComparator);
```
Why not 'reverseComparator' ?
'reverseOrder' is another possible name for this, as is 'reverseDefaultComparator'.

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -18,53 +18,321 @@
+  else if canResolveMethod(comparator, "key", data) {
+    // Check return type of key
+    type keytype = comparator.key(data).type;
+    if !(isNumericType(keytype) || isStringType(keytype)) {
```
The comments above indicate 'key must support numeric operations'. Should it say 'key must support numeric operations or be a string' ?

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -18,53 +18,321 @@
+/* Basic Functions */
+
+/*
+   Convenience wrapper around :proc:`quickSort`.
```
I'd prefer we described this function differently. Could it say

General purpose sorting interface. 

.. note:

  This method currently calls a sequential quickSort but it will call faster algorithms once they are implemented.

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -18,53 +18,321 @@
+ */
+proc isSorted(Data: [?Dom] ?eltType, comparator:?rec=defaultcomparator): bool {
+  chpl_check_comparator(comparator, eltType);
+  for i in Dom.low..Dom.high-1 do
```
I think it would be fun to write this in parallel but it doesn't need to happen right now...

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -18,53 +18,321 @@
+// array of x's element type" (or to infer its element type) without
+// saying a priori how big it is.  Without these mods, the result is
+// that the sorterator works when it does and probably is confusing
+// when it doesn't.
```
I'm confused by this comment. In the code below, y has element type inferred from x, and there is no mention of its size.

ben-albrecht replied:
Original comment by @bradcray, so he can correct me if I'm off:

I think it is saying that it would preferably create a new array `y` which shares the same element types `x`, but this requires a size to be passed (or we could calculate the size on the fly). This would be valuable for supporting iterable types other than arrays, such as iterators and lists.

bradcray replied:
Sorry that this wasn't clear on its own -- it's been long enough that I don't recall exactly what I was asking for, but my interpretation today is that I wanted a way to assert "y is an array" (where currently 'y' isn't constrained to be anything), but that I didn't have any way to do so since I couldn't predict its size.  This is akin to the longstanding desire to be able to say things like:

    var A: [1..] real = ...;

as a means of saying "It should be a 1-based array, but its size should be taken from the initialization expression.  We've discussed before how perhaps we should be able to declare variables that are first-class captures of iterators (like the David Wonnacott mail to chapel-bugs this week), and if we had those semantics, the current code wouldn't necessarily result in 'y' being array -- instead, it might be considered a capture of 'x' in the event that 'x' was, say, an iterator (or a non-array data structure that supported iteration?).  

I think the part about the element type is saying that it'd be nice if you could query arbitrary iterable things to ask what type you'll get when you iterate over them, so that one could say something like:

    var y: [1..] x.yieldType = for xs in x do (xs + 1);

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -18,53 +18,321 @@
+
+   :yields: The elements of x in sorted order
+   :ytype: x's element type
+
```
I'd add a .. note here that it's a TODO to make a standalone or leader/follower parallel iterator here.

mppf replied:
ie. that the current version is only a serial iterator.

ben-albrecht replied:
`TODO` + `.. note` added

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -283,64 +551,254 @@ proc QuickSort(Data: [?Dom] ?eltType, minlen=16, doublecheck=false, param revers
 
-  if (doublecheck) then VerifySort(Data, "SelectionSort", reverse);
+/* Reverse comparator built from another comparator.*/
+record ReverseComparator {
```
I wonder if this should just be called 'Reverse' ?

ben-albrecht replied:
I like the symmetry of `ReverseComparator` and `DefaultComparator`. I considered `Reverse` and `Default` with `reverse` and `default` instances, but these would conflict with variables in the deprecated functions and would rob users of some excellent variable names.

mppf commented:
@ben-albrecht - looks great. I had lots of little feedback, mostly about comments and names. Please let's discuss individual naming suggestions if it seems we are not in agreement.

Generally, I think there are some Future Work steps that I'd like you to call out in the PR/merge message:

* implement radix sort
* move individual sort algorithms to a SortAlgorithms module to reduce the complexity of Sort
   (since most uses of this library will only want to call 'sort', 'sorted' or 'isSorted').
* implement parallel sorting algorithms and have some reasonable way for users to request serial or parallel sorts.

Thanks!
